File: cpp/binary_search/search_a_2D_matrix_leetcode_74.cpp
#include <bits/stdc++.h>
#include <vector>
using namespace std;
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
      vector<int> valid_mat = find_vec(0, matrix.size() - 1, matrix, target);

      return(binary_search(0, valid_mat.size() - 1, valid_mat, target)); 
    }

    vector<int> find_vec(int start, int end, vector<vector<int>>& board, int target) {
      if (start > end){
        return {};
      }

      int mid = start + (end - start)/2;
      if(board[mid][0] <= target && board[mid][((board[mid]).size() - 1)] >= target){
        return board[mid];
      } else if (board[mid][0] > target){
        return find_vec(start, mid-1, board, target);
      } 
      return find_vec(mid+1, end, board, target);
    }

    bool binary_search(int start, int end, vector<int>& nums, int target){
      if(start > end){
        return false;
      }

      int mid = start + (end - start)/2;
      if(nums[mid] == target){
        return true;
      } else if (nums[mid] > target) {
       return binary_search(start, mid-1, nums, target);
      }
      return binary_search(mid+1, end, nums, target);
    }
};

int main(){
  Solution sol;
  std::vector<std::vector<int>> matrix = {
        {1, 3, 5, 7},
        {10, 11, 16, 20},
        {23, 30, 34, 60}
    };
  printf("{%d}\n", sol.searchMatrix(matrix, 10u));
}

File: python/binary_search/search_a_2D_matrix_leetcode_74.py
class Solution(object):
    def searchMatrix(self, matrix, target):
        valid_mat = self.find_vec(0, len(matrix) - 1, matrix, target)
        return self.binary_search(0, len(valid_mat) - 1, valid_mat, target)
    def find_vec(self, start, end, board, target):
        if start > end:
            return []
        mid = start + (end -  start) // 2
        l = board[mid][0]
        r = board[mid][len(board[mid])-1]
        if l <= target and r >= target:
            return board[mid]
        elif l > target:
            return self.find_vec(start, mid-1, board, target)
        return self.find_vec(mid+1, end, board, target)

    def binary_search(self, start, end, nums, target):
        if start > end:
            return False
        mid = start + (end - start) // 2
        if nums[mid] == target:
            return True
        elif nums[mid] > target:
            return self.binary_search(start, mid-1, nums, target)
        return self.binary_search(mid+1, end, nums, target)


if __name__ == "__main__":
    matrix = [[1,3,5,7],[10,11,16,20],[23,30,34,60]]
    sol = Solution()
    print(sol.searchMatrix(matrix, 3))
